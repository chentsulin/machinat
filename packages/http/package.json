{
  "name": "@machinat/http",
  "version": "0.4.0-beta.1",
  "author": "Patrick Liu <zeldalink0515@gmail.com>  (https://github.com/lrills)",
  "repository": {
    "type": "git",
    "url": "https://github.com/machinat/machinat.git",
    "directory": "packages/http"
  },
  "homepage": "https://github.com/machinat/machinat#readme",
  "bugs": "https://github.com/machinat/machinat/issues",
  "main": "lib/index.js",
  "exports": {
    ".": "./lib/index.js",
    "./types": "./lib/types.js",
    "./webhook": "./lib/webhook/index.js",
    "./webhook/types": "./lib/webhook/types.js"
  },
  "files": [
    "lib/*",
    "README.md",
    "polyfill-exports.js"
  ],
  "scripts": {
    "postinstall": "./polyfill-exports.js"
  },
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "invariant": "^2.2.4",
    "raw-body": "^2.4.1",
    "thenifiedly": "^0.2.1"
  },
  "peerDependencies": {
    "@machinat/core": "^0.4.1-alpha"
  }
}
