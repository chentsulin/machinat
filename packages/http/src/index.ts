export { default } from './module';
export { HttpConnector } from './connector';
export * from './types';
