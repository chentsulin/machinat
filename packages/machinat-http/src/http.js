// @flow
import url from 'url';
import type { IncomingMessage, ServerResponse } from 'http';
import { HTTPReceivable } from './types';

export const connect = (bot: HTTPReceivable) => (
  req: IncomingMessage,
  res: ServerResponse
) => {
  bot.adaptor.handleRequest(req, res);
};

export const connectByRoutes = (mapping: { [string]: HTTPReceivable }) => {
  const routes = Object.keys(mapping);

  return (req: IncomingMessage, res: ServerResponse) => {
    const parsed = url.parse(req.url);
    const pathname = parsed.pathname || '';

    for (const route of routes) {
      if (pathname.slice(-route.length) === route) {
        mapping[route].adaptor.handleRequest(req, res);
        return;
      }
    }

    res.statusCode = 404; // eslint-disable-line no-param-reassign
    res.end();
  };
};
