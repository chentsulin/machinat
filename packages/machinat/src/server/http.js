import http from 'http';

const createServer = connector => {
  const handleWebhook = connector.handleWebhook();
  const empty = () => {};

  return http.createServer((req, res) => {
    handleWebhook(req, res).catch(empty);
  });
};

export default createServer;
