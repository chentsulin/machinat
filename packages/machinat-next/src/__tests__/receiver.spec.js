import { IncomingMessage, ServerResponse } from 'http';
import moxy from 'moxy';
import NextReceiver from '../receiver';

const nextTick = () => new Promise(resolve => process.nextTick(resolve));

const next = moxy({
  render: async () => {},
  renderError: async () => {},
});

const req = moxy(new IncomingMessage());
const res = moxy(new ServerResponse({ method: 'GET' }));
req.mock.getter('method').fakeReturnValue('GET');
req.mock.getter('url').fakeReturnValue('http://machinat.com/hello?foo=bar');
req.mock.getter('headers').fakeReturnValue({ 'x-y-z': 'abc' });
res.mock.setter('statusCode').fake(() => {});

const issueEvent = moxy(async () => ({
  pathname: '/hello',
  query: { foo: 'bar' },
}));

const issueError = moxy();

beforeEach(() => {
  next.mock.clear();
  req.mock.clear();
  res.mock.clear();

  issueEvent.mock.reset();
  issueError.mock.reset();
});

it('renderError with 501 if receiver is not bound', async () => {
  const receiver = new NextReceiver(next);

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(next.renderError.mock).toHaveBeenCalledTimes(1);
  expect(next.renderError.mock).toHaveBeenCalledWith(null, req, res, '/hello', {
    foo: 'bar',
  });

  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(501);
});

it('render', async () => {
  const receiver = new NextReceiver(next);
  receiver.bindIssuer(issueEvent, issueError);

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(next.render.mock).toHaveBeenCalledTimes(1);
  expect(next.render.mock).toHaveBeenCalledWith(req, res, '/hello', {
    foo: 'bar',
  });

  expect(issueEvent.mock).toHaveBeenCalledTimes(1);
  expect(issueEvent.mock).toHaveBeenCalledWith(
    {
      platform: 'next',
      type: 'server',
      uid: 'next:server',
    },
    {
      platform: 'next',
      type: 'request',
      payload: {
        pathname: '/hello',
        query: { foo: 'bar' },
      },
    },
    {
      source: 'next',
      request: {
        method: 'GET',
        url: 'http://machinat.com/hello?foo=bar',
        headers: { 'x-y-z': 'abc' },
      },
    }
  );

  expect(res.mock.setter('statusCode')).not.toHaveBeenCalled();
});

it('render with params return by event issuer', async () => {
  const receiver = new NextReceiver(next);
  receiver.bindIssuer(issueEvent, issueError);

  issueEvent.mock.fake(async () => ({
    pathname: '/hello/world',
    query: { foo: 'baz' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(issueEvent.mock).toHaveBeenCalledTimes(1);
  expect(res.mock.setter('statusCode')).not.toHaveBeenCalled();

  expect(next.render.mock).toHaveBeenCalledTimes(1);
  expect(next.render.mock).toHaveBeenCalledWith(req, res, '/hello/world', {
    foo: 'baz',
  });
});

it('renderError renderError with 501 if event issuer return empty', async () => {
  const receiver = new NextReceiver(next);
  receiver.bindIssuer(issueEvent, issueError);

  issueEvent.mock.fake(async () => {});

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(issueEvent.mock).toHaveBeenCalledTimes(1);
  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(501);

  expect(next.render.mock).not.toHaveBeenCalled();
  expect(next.renderError.mock).toHaveBeenCalledTimes(1);
  expect(next.renderError.mock).toHaveBeenCalledWith(null, req, res, '/hello', {
    foo: 'bar',
  });
});

it('renderError if event issuer thrown', async () => {
  const receiver = new NextReceiver(next);
  receiver.bindIssuer(issueEvent, issueError);

  issueEvent.mock.fake(async () => {
    throw new Error("I'm a teapot");
  });

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(issueEvent.mock).toHaveBeenCalledTimes(1);

  expect(next.render.mock).not.toHaveBeenCalled();
  expect(next.renderError.mock).toHaveBeenCalledTimes(1);
  expect(next.renderError.mock).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(issueError.mock).toHaveBeenCalledTimes(1);
  expect(issueError.mock).toHaveBeenCalledWith(new Error("I'm a teapot"));
});
