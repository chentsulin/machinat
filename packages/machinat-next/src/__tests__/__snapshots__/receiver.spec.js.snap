// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call next.render() if there is cutomized page or query return by middlewares 1`] = `
Object {
  "bot": null,
  "channel": Object {
    "platform": "next",
    "type": "server",
    "uid": "next.server",
  },
  "event": Object {
    "kind": "request",
    "payload": Object {
      "request": Object {
        "headers": Object {
          "x-y-z": "abc",
        },
        "method": "GET",
        "url": "http://machinat.com/hello?foo=bar",
      },
    },
    "platform": "next",
    "type": "request",
  },
  "metadata": Object {
    "request": Object {
      "headers": Object {
        "x-y-z": "abc",
      },
      "method": "GET",
      "url": "http://machinat.com/hello?foo=bar",
    },
    "source": "next",
  },
  "platform": "next",
  "user": null,
}
`;

exports[`call next.renderError() if middlewares resolve unaccepted 1`] = `
Object {
  "bot": null,
  "channel": Object {
    "platform": "next",
    "type": "server",
    "uid": "next.server",
  },
  "event": Object {
    "kind": "request",
    "payload": Object {
      "request": Object {
        "headers": Object {
          "x-y-z": "abc",
        },
        "method": "GET",
        "url": "http://machinat.com/hello?foo=bar",
      },
    },
    "platform": "next",
    "type": "request",
  },
  "metadata": Object {
    "request": Object {
      "headers": Object {
        "x-y-z": "abc",
      },
      "method": "GET",
      "url": "http://machinat.com/hello?foo=bar",
    },
    "source": "next",
  },
  "platform": "next",
  "user": null,
}
`;

exports[`set customized headers return by middlewares on res 1`] = `
Object {
  "bot": null,
  "channel": Object {
    "platform": "next",
    "type": "server",
    "uid": "next.server",
  },
  "event": Object {
    "kind": "request",
    "payload": Object {
      "request": Object {
        "headers": Object {
          "x-y-z": "abc",
        },
        "method": "GET",
        "url": "http://machinat.com/hello?foo=bar",
      },
    },
    "platform": "next",
    "type": "request",
  },
  "metadata": Object {
    "request": Object {
      "headers": Object {
        "x-y-z": "abc",
      },
      "method": "GET",
      "url": "http://machinat.com/hello?foo=bar",
    },
    "source": "next",
  },
  "platform": "next",
  "user": null,
}
`;

exports[`trim the entryPath from the url passed to next handler 1`] = `
Object {
  "bot": null,
  "channel": Object {
    "platform": "next",
    "type": "server",
    "uid": "next.server",
  },
  "event": Object {
    "kind": "request",
    "payload": Object {
      "request": Object {
        "headers": Object {
          "x-y-z": "abc",
        },
        "method": "GET",
        "url": "http://machinat.com/hello/world",
      },
    },
    "platform": "next",
    "type": "request",
  },
  "metadata": Object {
    "request": Object {
      "headers": Object {
        "x-y-z": "abc",
      },
      "method": "GET",
      "url": "http://machinat.com/hello/world",
    },
    "source": "next",
  },
  "platform": "next",
  "user": null,
}
`;
